generate/src/contract/events.rs

```diff
@@ -1,21 +1,189 @@
-use crate::contract::{types, Context};
+use crate::contract::{methods, types, Context};
 use crate::util;
 use anyhow::Result;
 use ethcontract_common::abi::{Event, Hash};
+use ethcontract_common::abiext::EventExt;
 use inflector::Inflector;
 use proc_macro2::{Literal, TokenStream};
 use quote::quote;
 
 pub(crate) fn expand(cx: &Context) -> Result<TokenStream> {
+    let structs_mod = expand_structs_mod(cx)?;
     let filters = expand_filters(cx)?;
     let all_events = expand_all_events(cx);
 
     Ok(quote! {
+        #structs_mod
         #filters
         #all_events
     })
 }
 
+/// Expands into a module containing all the event data structures from the ABI.
+fn expand_structs_mod(cx: &Context) -> Result<TokenStream> {
+    let data_types = cx
+        .artifact
+        .abi
+        .events()
+        .map(|event| expand_data_type(event))
+        .collect::<Result<Vec<_>>>()?;
+    if data_types.is_empty() {
+        return Ok(quote! {});
+    }
+
+    Ok(quote! {
+        /// Module containing all generated data models for this contract's
+        /// events.
+        pub mod events {
+            use super::ethcontract;
+
+            #( #data_types )*
+        }
+    })
+}
+
+/// Expands an ABI event into a single event data type. This can expand either
+/// into a structure or a tuple in the case where all event parameters (topics
+/// and data) are anonymous.
+fn expand_data_type(event: &Event) -> Result<TokenStream> {
+    let event_name = expand_struct_name(event);
+
+    let signature = expand_hash(event.signature());
+    let abi_signature = Literal::string(&event.abi_signature());
+
+    let params = expand_params(event)?;
+
+    let all_anonymous_fields = event.inputs.iter().all(|input| input.name.is_empty());
+    let (data_type_def, data_type_cstr) = if all_anonymous_fields {
```

Comment:
ubernit: maybe data_type_definition/construction instead of abbreviation (cstr reminds me of C++)

Reply:
Hehe